### PR TITLE
Fedmeta

### PR DESCRIPTION
**Pull request by jinsoogod**
### Description

Implements [FedMeta](https://arxiv.org/abs/1802.07876) baseline.

### Related issues/PRs

See [issue 2241](https://github.com/adap/flower/issues/2241)

### Checklist

- [ ] Implement proposed change
- [ ] Write tests
- [ ] Update [documentation](https://flower.dev/docs/writing-documentation.html)
- [ ] Update [changelog](https://github.com/adap/flower/blob/main/doc/source/changelog.rst)
- [ ] Make CI checks pass
- [ ] Ping maintainers on [Slack](https://flower.dev/join-slack/) (channel `#contributions`)

### CODE REVIEW

File: baselines/fedmeta/fedmeta/conf/data/shakespeare.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+# this is the config that will be loaded as default by main.py
+# Please follow the provided structure (this will ensuring all baseline follow
+# a similar configuration structure and hence be easy to customise)
+
+model:
+  _target_: models.StackedLSTM
+
+client_resources:
+  num_cpus: 4
+  num_gpus: 1.0
```
Comment (jafermarq) [inline baselines/fedmeta/fedmeta/conf/data/shakespeare.yaml:11]:
```suggestion
  num_cpus: 4
  num_gpus: 0.2
```

Comment (jafermarq) [reply baselines/fedmeta/fedmeta/conf/data/shakespeare.yaml:11]: I think these clients do not need an entire GPU (VRAM usage < 1GB). Or did you experience some problems using a `num_gpus` lower than 1?

File: baselines/fedmeta/README.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+|   FedAvg(Meta)    |     FEMNST     |         4         |       2000       |     10     |   Adam    |       0.0001        | {'num_cpus': 4.0, 'num_gpus': 0.25 } |       -       |
+|   FedAvg(Meta)    |  SHAKESPEARE   |         4         |       400        |     10     |   Adam    |        0.001        | {'num_cpus': 4.0, 'num_gpus': 0.25 } |       -       |
+|   FedMeta(MAML)   |     FEMNST     |         4         |       2000       |     10     |   Adam    |   (0.001, 0.0001)   | {'num_cpus': 4.0, 'num_gpus': 1.0 }  |       5       |
+|   FedMeta(MAML)   |  SHAKESPEARE   |         4         |       400        |     10     |   Adam    |     (0.1, 0.01)     | {'num_cpus': 4.0, 'num_gpus': 1.0 }  |       1       |
+| FedMeta(Meta-SGD  |     FEMNST     |         4         |       2000       |     10     |   Adam    |   (0.001, 0.0001)   | {'num_cpus': 4.0, 'num_gpus': 1.0 }  |       5       |
+| FedMeta(Meta-SGD  |  SHAKESPEARE   |         4         |       400        |     10     |   Adam    |     (0.1, 0.01)     | {'num_cpus': 4.0, 'num_gpus': 1.0 }  |       1       |
```
Comment (jafermarq) [inline baselines/fedmeta/README.md:60]:
```suggestion
|      FedAvg       |     FEMNIST<br>SHAKESPEARE     |         4         |       2000<br>400       |     10     |   Adam    |       0.0001<br>0.001        | {'num_cpus': 4.0, 'num_gpus': 0.25 } |       -       |
|   FedAvg(Meta)    |     FEMNIST<br>SHAKESPEARE     |         4         |       2000<br>400       |     10     |   Adam    |       0.0001<br>0.001        | {'num_cpus': 4.0, 'num_gpus': 0.25 } |       -       |
|   FedMeta(MAML)   |     FEMNIST<br>SHAKESPEARE     |         4         |       2000<br>400       |     10     |   Adam    |   (0.001, 0.0001)<br>(0.1, 0.01)   | {'num_cpus': 4.0, 'num_gpus': 1.0 }  |       5<br>1       |
| FedMeta(Meta-SGD)  |     FEMNIST<br>SHAKESPEARE     |         4         |       2000<br>400       |     10     |   Adam    |   (0.001, 0.0001)<br>(0.1, 0.01)   | {'num_cpus': 4.0, 'num_gpus': 1.0 }  |       5<br>1       |
```

Comment (jafermarq) [reply baselines/fedmeta/README.md:60]: Using `<br>` makes it look nicer. Up to you to use this

File: baselines/fedmeta/README.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+Pip install torch torch vision
+Pip install matplotlib
+pip install scikit-learn
+```
+
+## Running the Experiments
+
+****Download Dataset**** : Go [LEAF: A Benchmark for Federated Settings](https://github.com/TalwalkarLab/leaf) and Use the command below! You can download dataset (FEMNIST and SHAKESPEARE). 
+```bash
+#FEMNIST dataset Download command for these experiments
```
Comment (jafermarq) [inline baselines/fedmeta/README.md:77]: I'm thinking you could explain a bit what each of those flags passed to `./preprocess.sh` briefly do

File: baselines/fedmeta/README.md [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+./preprocess.sh -s niid --iu 1068 --sf 0.3 -k 0 -t sample
+
+#SHAKESEPEARE dataset Download command for these experiments
+./preprocess.sh -s niid --sf 0.16 -k 0 -t sample
+```
+
+****Start experiments**** : 
+```bash  
+# FedAvg + Femnist Dataset
+python main.py algo=fedavg data=femnist path=(your leaf dataset path)/leaf/data/shakespeare/data
```
Comment (jafermarq) [inline baselines/fedmeta/README.md:87]:
```suggestion
python -m fedmeta.main algo=fedavg data=femnist path=(your leaf dataset path)/leaf/data/shakespeare/data
```

File: baselines/fedmeta/README.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+```
+
+
+## Expected Results
+If you proceed with all of the above experiments, You can get a graph of your experiment results as shown below along that `./femnist or shakespeare/graph_params/result_graph.png`.
+
+**Femnist dataset experiment results**
+
+
+![](docs/femnist_result_graph.png)
```
Comment (jafermarq) [inline baselines/fedmeta/README.md:121]: Could we have grid lines for this and the shakespeare plots? also, rename the directory to `_static`

**Review by jafermarq — COMMENTED** [review]
Hi @jinsoogod ,

I have completed a first pass for your `FedMeta` implementation. It looks good!!

You'll see i've done several suggestions to the code and the configs. Maybe of them are related and are very minor. In summary, my suggestions:
* are done so to treat `fedmeta` as a proper python package (that's built w/ poetry). This means that the path to your modules needs the `fedmeta.` prefix. You'll need to change these in the `.py` files and in the configs (`_target_` elements). I was able to run the code just fine after applying all those changes.
* Now all packages needed (e.g. torch, matplotlib, etc) are in `pyproject.toml`, so everything is installed via `poetry install`. I have extended a bit the instructions to build the python environment.
* I made some suggestions on how to others interested in your baseline to use `LEAF` dataset.
* Other small changes to the `README.md`.

Also, could you please remove the changes to files outside your `fedmeta` directory in `/baselines`? (for example, those in `doc/`, `examples/`, etc)

Please ping me when you are done, don't forget to mark as resolved the items below once you made the fix. Happy to discuss about them further if you don't agree with some of my suggestions.

File: baselines/fedmeta/fedmeta/conf/data/shakespeare.yaml (hunk 1 above)
Comment (jinsoogod) [reply baselines/fedmeta/fedmeta/conf/data/shakespeare.yaml:11]: Hi, @jafermarq. For both the Femnist and Shakespeare datasets, basic federated learning (FedAvg and FedAvg_Meta) can be trained using less than one GPU. However, during my experiment, when meta-learning was incorporated, many clients experienced out-of-memory errors during the second derivative computation of the LSTM model.

File: baselines/fedmeta/README.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+Poetry install
+Poetry shell
+Pip install torch torch vision
+Pip install matplotlib
+pip install scikit-learn
+```
+
+## Running the Experiments
+
+****Download Dataset**** : Go [LEAF: A Benchmark for Federated Settings](https://github.com/TalwalkarLab/leaf) and Use the command below! You can download dataset (FEMNIST and SHAKESPEARE).
```
[the remark this answers is not in the document]
Comment (jinsoogod) [reply baselines/fedmeta/README.md:76]: I agree with you. I'll try to write a more detailed explanation.

File: baselines/fedmeta/README.md (hunk 4 above)
Comment (jinsoogod) [reply baselines/fedmeta/README.md:88]: The code has been adjusted to fit the command.

**Comment by jinsoogod** [comment]
Hi, @jafermarq. I have made several revisions based on your suggestions. Below are the answers to your questions.

Q.I think these clients do not need an entire GPU (VRAM usage < 1GB). Or did you experience some problems using a num_gpus lower than 1?

> For both the Femnist and Shakespeare datasets, basic federated learning (FedAvg and FedAvg_Meta) can be trained using less than one GPU. However, during my experiment, when meta-learning was incorporated, many clients experienced out-of-memory errors during the second derivative computation of the LSTM model.

If you have any further suggestions, please don't hesitate to let me know. I'm always open to feedback!

File: baselines/fedmeta/pyproject.toml [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    "Topic :: Software Development :: Libraries :: Python Modules",
+    "Typing :: Typed",
+]
+
+[tool.poetry.dependencies]
+python = ">=3.10.0, <3.11.0" 
+flwr = { extras = ["simulation"], version = "1.5.0" }
+hydra-core = "1.3.2" # don't change this
+matplotlib = "3.7.1"
+scikit-learn = "^1.3.1"
```
Comment (jafermarq) [inline baselines/fedmeta/pyproject.toml:44]:
```suggestion
scikit-learn = "1.3.1"
```

Comment (jafermarq) [reply baselines/fedmeta/pyproject.toml:44]: let's just fix it

File: baselines/fedmeta/README.md [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+**Shakespeare Dataset Issue** : In the FedMeta paper experiment, the Shakespeare dataset had 1126 users. However, due to a current bug, the number of users has decreased to 660 users. Therefore, we have only maintained the total number of data.
+
+|   Dataset   |      #Clients       |     #Samples     | #Classes |                       #Partition Clients                        |   #Partition Dataset   |
+|:-----------:|:-------------------:|:----------------:|:--------:|:---------------------------------------------------------------:|:----------------------:|
+|   FEMNIST   |    About<br>1050    | About<br>240,000 |    62    | Train Clients : 0.8 <br>Valid Clients : 0.1, Test Clients : 0.1 | Sup : 0.2<br>Qry : 0.8 |
+| SHAKESPEARE | About<br>550 -> 110 | About<br>620,000 |    80    | Train Clients : 0.8 <br>Valid Clients : 0.1, Test Clients : 0.1 |   Sup : 0.2<br>Qry : 0.8   |
```
Comment (jafermarq) [inline baselines/fedmeta/README.md:45]: Could we get the exact number for `#Samples`?

File: baselines/fedmeta/README.md [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+```
+
+
+## Expected Results
+If you proceed with all of the above experiments, You can get a graph of your experiment results as shown below along that `./femnist or shakespeare/graph_params/result_graph.png`.
+
+|                   FEMNIST                   |                     SHAKESPEARE                      |
+|:-------------------------------------------:|:----------------------------------------------------:|
+|    ![](_static/femnist_result_graph.png)    |      ![](_static/shakespeare_result_graph.png)       |
```
Comment (jafermarq) [inline baselines/fedmeta/README.md:139]: Your plots look very good. Would it be possible to have grid lines in them?

**Review by jafermarq — COMMENTED** [review]
Hi @jinsoogod,

Thanks for making the changes suggested in my previous review. I have listed a few small suggestions. I think your baseline looks in a very good state already!

Please don't forget to run `./dev/format-baseline.sh fedmeta` and ensure you don't see any error/warning. These are generally easy to fix. Please let me know if you find some harder to resolve, I can help. Once the command doesn't show any more errors, please run `./dev/test-baseline.sh fedmeta` and fix errors if they appear.

These formatting and testing checks might sound trivial but are necessary for the baseline to get merged.

**Comment by jinsoogod** [comment]
Hi, @jafermarq. I've completed the edits based on your second request. I've addressed most of the execution errors from `./dev/format-baseline.sh fedmeta` and `./dev/test-baseline.sh fedmeta`. I'd appreciate feedback for further improvements.

File: baselines/fedmeta/fedmeta/femnist/graph_params/result_graph.png
Comment (jafermarq) [inline baselines/fedmeta/fedmeta/femnist/graph_params/result_graph.png:1]: please remove this, it's already in `_static`.

File: baselines/fedmeta/pyproject.toml [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    "Typing :: Typed",
+]
+
+[tool.poetry.dependencies]
+python = ">=3.10.0, <3.11.0" 
+flwr = { extras = ["simulation"], version = "1.5.0" }
+hydra-core = "1.3.2" # don't change this
+matplotlib = "3.7.1"
+scikit-learn = "1.3.1"
+
```
Comment (jafermarq) [inline baselines/fedmeta/pyproject.toml:45]:
```suggestion
torch = { url = "https://download.pytorch.org/whl/cu117/torch-2.0.1%2Bcu117-cp310-cp310-linux_x86_64.whl"}
torchvision = { url = "https://download.pytorch.org/whl/cu117/torchvision-0.15.2%2Bcu117-cp310-cp310-linux_x86_64.whl"}


```

Comment (jafermarq) [reply baselines/fedmeta/pyproject.toml:45]: Let's add `torch` and `torchvision` here so we don't need to do `poetry add ...` after doing `poetry install`. I've adjusted your `README.md` in another comment.

File: baselines/fedmeta/README.md [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+#Environment Setup
+# Set python version
+pyenv install 3.10.6
+pyenv local 3.10.6
+
+# Tell poetry to use python 3.10
+poetry env use 3.10.6
+
+# install the base Poetry environment
+poetry add torch torchvision
```
[the remark this answers is not in the document]
Comment (jafermarq) [reply baselines/fedmeta/README.md:70]: remove since we install this in `pyproject.toml`.

File: baselines/fedmeta/pyproject.toml (hunk 10 above)
Comment (jafermarq) [reply baselines/fedmeta/pyproject.toml:45]: If you prefer having a different version, please edit the above.

File: baselines/fedmeta/fedmeta/main.py [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    # Hydra will generate for you a directory each time you run the code. You
+    # can retrieve the path to that directory with this:
+    # save_path = HydraConfig.get().runtime.output_dir
+
+    print("................")
+    print(history)
+    output_path = (
+        HydraConfig.get().runtime.cwd + "/fedmeta/" + cfg.data.data + "/graph_params"
+    )
+    os.makedirs(output_path, exist_ok=True)
```
[the remark this answers is not in the document]
Comment (jafermarq) [reply baselines/fedmeta/fedmeta/main.py:90]: Wouldn't it be better to save the images to the directory in `outputs/` that Hydra automatically generates for you? (there is where the log and config gets copied). I think it makes more sense to get the plots generated and saved there.

[the remark this answers is not in the document]
Comment (jafermarq) [reply baselines/fedmeta/fedmeta/main.py:90]: I have tested this, and it works as expected.

File: baselines/fedmeta/fedmeta/client.py [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+    Returns
+    -------
+    Tuple[Callable[[str], FlowerClient], DataLoader]
+        A tuple containing the client function that creates Flower Clients and
+        the DataLoader that will be used for testing
+    """
+
+    def client_fn(cid: str) -> FlowerClient:
+        """Create a Flower client representing a single organization."""
+        print(f"cid : {cid}")
```
[the remark this answers is not in the document]
Comment (jafermarq) [reply baselines/fedmeta/fedmeta/client.py:168]: Let's remove this?

File: baselines/fedmeta/pyproject.toml [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+[tool.mypy]
+ignore_missing_imports = true
+strict = false
+plugins = "numpy.typing.mypy_plugin"
+
+[tool.pylint."MESSAGES CONTROL"]
+disable = "bad-continuation,duplicate-code,too-few-public-methods,useless-import-alias"
+good-names = "i,j,k,_,x,y,X,Y"
+signature-mutators="hydra.main.main"
+
```
[the remark this answers is not in the document]
Comment (jafermarq) [reply baselines/fedmeta/pyproject.toml:86]: this was added to the template `pyproject.toml` after you created your baseline. this is needed, so please include it.

**Review by jafermarq — COMMENTED** [review]
Hi @jinsoogod,

Thanks for including all those changes. I have listed a few more below. In summary:

* Your PR contains changes to files outside `baselines/fedmeta` those need to be discarded. I can help with this before merging if it gets complicated
* Remove the plots in `<dataset-name>/graph_params` since they are in `_static` already
* Edit the `main.py` so the plots are saved to the directory of the experiment automatically. Then probably you want to indicate at the end of your readme, that the location of the plots is the output directory of the experiment.
* Install `torch` and `torchvision` directly with `poetry install` (so those are moved to `pyproject.toml`)
* Other small changes

I can confirm that your tests run fine, but you need to add a few extra lines to your `pyproject.toml` that i also indicate below.

Please ping me once you had a chance to look into everything I mention. Your baseline looks good!

**Comment by jinsoogod** [comment]
@jafermarq, I have completed the list you suggested. Please let me know what further steps I need to take.

File: baselines/fedmeta/README.md [hunk 15]
```diff
@@ -0,0 +1,139 @@
+---
+title: Federated Meta-Learning with Fast Convergence and Efficient Communication
+url: https://arxiv.org/abs/1802.07876
+labels: [meta learning, maml, meta-sgd, personalization] 
+dataset: [FEMNIST, SHAKESPEARE]
+---
+
+# FedMeta: Federated Meta-Learning with Fast Convergence and Efficient Communication
+
+****Paper:**** : [arxiv.org/abs/1802.07876](https://arxiv.org/abs/1802.07876)
```
Comment (jafermarq) [inline baselines/fedmeta/README.md:10]:
```suggestion
**Paper:** [arxiv.org/abs/1802.07876](https://arxiv.org/abs/1802.07876)
```

**Review by jafermarq — COMMENTED** [review]
@jinsoogod , thanks for making those changes. You'll notice removed some files that were not part of your `baseline/fedmeta` (there are a couple more that i need to look into -- not sure what's going on exactly). 

Could you in the meantime:
* Run the `format-baseline.sh` script again? some things need to change (this will happen automatically)
* make the suggested changes to the readme.
Thank you.

**Comment by jinsoogod** [comment]
@jafermarq , I've fixed your edits now. Please check

**Comment by jafermarq** [comment]
@jinsoogod, I have added some updates to fix the formatting and also discard files that weren't part of the baseline. Until now i've only tested your code with Shakespeare (which i'm running now for all 400 rounds). I'm running FEMNIST for the first time and I noticed some errors in the `LEAF` preprocessing script. See below:

```
hashed 700000 write images
hashed 800000 write images
finished calculating image hashes
------------------------------
assigning class labels to write images
finished assigning class labels to write images
------------------------------
grouping images by writer
finished grouping images by writer
------------------------------
converting data to .json format
Traceback (most recent call last):
  File "/home/javier/projects/sor/fedmeta/baselines/fedmeta/leaf/data/femnist/preprocess/data_to_json.py", line 64, in <module>
    gray.thumbnail(size, Image.ANTIALIAS)
AttributeError: module 'PIL.Image' has no attribute 'ANTIALIAS'
finished converting data to .json format
------------------------------

[THEN THE REST FAILS ALSO]

```

Could you:
* look into this errors? It seems these can be fixed if downgrading `Pillow` to 9.5, but i'm not sure if this will affect the results you already obtained. What version of Pillow did you have in your environment when running the code? `Poetry` installs Pillow when doing `poetry install` and uses version `10.0.1`

**Comment by jinsoogod** [comment]
@jafermarq, You are right. LEAF github was created with pillow version 9.5. I downloaded this data from the existing LEAF Anaconda environment. The current error occurs during the download process. This does not occur during preprocessing that maintains raw data. My Pillow version is 9.5.0

**Comment by jafermarq** [comment]
Ok, i'll add that as part of `pyproject.toml`. I'll re-run some of your experiments and if all looks good, we are ready to merge!

**Comment by jinsoogod** [comment]
Thank you for your hard work. I'm waiting for good news.

**Review by jafermarq — APPROVED** [review]
Looks great! @jinsoogod

**Review by danieljanes — APPROVED** [review]

**Pull request merged**